geode-core/src/main/java/org/apache/geode/internal/cache/wan/AbstractGatewaySender.java

```diff
@@ -1118,6 +1115,17 @@ public void distribute(EnumListenerEvent operation, EntryEventImpl event,
     }
   }
 
+  private void recordDroppedEvent(EntryEventImpl event) {
+    if (this.eventProcessor != null) {
+      this.eventProcessor.registerEventDroppedInPrimaryQueue(event);
```

Comment:
In ConcurrentParallelGatewaySenderEventProcessor.registerEventDroppedInPrimaryQueue instead of calling sendQueueRemovalMesssageForDroppedEvent to send a ParallelQueueRemovalMessage inline, can you call ParallelGatewaySenderQueue.addRemovedEvent and let the BatchRemovalThread handle sending the message?

Something like:
```
...
int bucketId = PartitionedRegionHelper.getHashKey((EntryOperation) droppedEvent);
boolean isPrimary = prQ.getRegionAdvisor().getBucketAdvisor(bucketId).isPrimary();
if (isPrimary) {
  long shadowKey = droppedEvent.getTailKey();
  ParallelGatewaySenderQueue pgsq = (ParallelGatewaySenderQueue) cpgsq.getQueueByBucket(bucketId);
  pgsq.addRemovedEvent(prQ, bucketId, shadowKey);
  this.sender.getStatistics().incEventsDroppedDueToPrimarySenderNotRunning();
...
``
I didn't try this, but it seems like it should work.

I'm not sure if the same idea would work for SerialGatewaySenderEventProcessor.sendBatchDestroyOperationForDroppedEvent since it has a different implementation.

btw - sendQueueRemovalMesssageForDroppedEvent has an extra 's'.

Reply:
@boglesby I have made the change you have proposed for the parallel gateway sender case. You can check it in the new commit. Nevertheless, this change it requires that the BatchRemoval thread is not shutdown when the gateway sender is stopped. Is there any problem with not stopping the BatchRemoval thread when the sender is stopped?

Regarding the SerialGatewaySender, I think I found a way of doing it which consists of changing the following call in SerialGatewaySenderEventProcessor.registerEventDroppedInPrimaryQueue():

``
this.processors.get(index).sendBatchDestroyOperationForDroppedEvent(droppedEvent, index);
``
to this one:

``
this.processors.get(index).sendBatchDestroyOperationForDroppedEvent(droppedEvent, index);
``

And adding the following method to SerialGatewaySenderQueue:
``
  public void addRemovedEvent(EntryEventImpl droppedEvent) {
    lock.writeLock().lock();
    lastDispatchedKey = droppedEvent.getTailKey();
    notifyAll();
    lock.writeLock().unlock();
  }
``

The problem is that I did not find any test case to verify that it worked correctly. I ran the tests in SerialWANPropagationsFeatureDUnitTest and it did not matter if I used the original solution, the new solution or even removed the sending of batch destroy operation. I looks as if it were not necessary. That's why I did not dare to add the change. Any idea?

Reply:
Oh, I didn't think about the BatchRemoval thread not being there. I'm not really sure the consequences of keeping that thread around. I just seems better for performance to not send a message on every operation, and instead batch them up. Maybe the behavior can be something like: If that cache is closing, stop the BRM; otherwise keep it around to process these messages.

Reply:
How could I know if the cache is closing from inside the stop method of the ParallelGatewaySender? Is there a way?

Reply:
@boglesby I reverted the change about letting the batch removal thread remove the dropped events because I did not find a way to stop the thread when the cache was closing.

@boglesby and @gesterzhou , could you please approve the PR if you are ok with the changes? Otherwise, please let me know what else should be changed.

Reply:
@boglesby and @gesterzhou, friendly reminder :-)